README.pod
=pod

=head1 NAME

Dancer2::Plugin::Auth::Tiny - Require logged-in user for specified routes

=head1 VERSION

version 0.003

=head1 SYNOPSIS

  use Dancer2::Plugin::Auth::Tiny;

  get '/private' => needs login => sub { ... };

  get '/login' => sub {
    # put 'return_url' in a hidden form field
    template 'login' => { return_url => params->{return_url} };
  };

  post '/login' => sub {
    if ( _is_valid( params->{user}, params->{password} ) ) {
      session user => params->{user},
      return redirect params->{return_url} || '/';
    }
    else {
      template 'login' => { error => "invalid username or password" };
    }
  };

  sub _is_valid { ... } # this is up to you

=head1 DESCRIPTION

This L<Dancer2> plugin provides an extremely simple way of requiring that a user
be logged in before allowing access to certain routes.

It is not "Tiny" in the usual CPAN sense, but it is "Tiny" with respect to
Dancer2 authentication plugins.  It provides very simple sugar to wrap route
handlers with an authentication closure.

The plugin provides the C<needs> keyword and a default C<login> wrapper that
you can use like this:

  get '/private' => needs login => $coderef;

The code above is roughly equivalent to this:

  get '/private' => sub {
    if ( session 'user' ) {
      goto $coderef;
    }
    else {
      return redirect uri_for( '/login',
        { return_url => uri_for( request->path, request->params ) } );
    }
  };

It is up to you to provide the '/login' route, handle actual authentication,
and set C<user> session variable if login is successful.

If the original request contains a parameter in the C<passthrough> list, it
will be added to the login query. For example,
C<http://example.com/private?user=dagolden> will be redirected as
C<http://example.com/login?user=dagolden&return_url=...>.  This facilitates
pre-populating a login form.

=for Pod::Coverage extend

=head1 CONFIGURATION

You may override any of these settings:

=over 4

=item *

C<login_route: /login> -- defines where a protected route is redirected

=item *

C<logged_in_key: user> -- defines the session key that must be true to indicate a logged-in user

=item *

C<callback_key: return_url> -- defines the parameter key with the original request URL that is passed to the login route

=item *

C<passthrough: - user> -- a list of parameters that should be passed through to the login handler

=back

=head1 EXTENDING

The class method C<extend> may be used to add (or override) authentication
criteria. For example, to add a check for the C<session 'is_admin'> key:

  Dancer2::Plugin::Auth::Tiny->extend(
    admin => sub {
      my ($dsl, $coderef) = @_;
      return sub {
        if ( $dsl->app->session("is_admin") ) {
          goto $coderef;
        }
        else {
          $dsl->app->context->redirect '/access_denied';
        }
      };
    }
  );

  get '/super_secret' => needs admin => sub { ... };

It takes key/value pairs where the value must be a closure generator that wraps
arguments passed to C<needs>.

You could pass additional arguments before the code reference like so:

  # don't conflict with Dancer2's any()
  use Syntax::Keyword::Junction 'any' => { -as => 'any_of' };

  Dancer2::Plugin::Auth::Tiny->extend(
    any_role => sub {
      my $coderef = pop;
      my ($dsl, @requested_roles) = @_;
      return sub {
        my @user_roles = @{ $dsl->app->session("roles") || [] };
        if ( any_of(@requested_roles) eq any_of(@user_roles) ) {
          goto $coderef;
        }
        else {
          $dsl->app->context->redirect '/access_denied';
        }
      };
    }
  );

  get '/parental' => needs any_role => qw/mom dad/ => sub { ... };

=head1 SEE ALSO

For more complex L<Dancer2> authentication, see:

=over 4

=item *

L<Dancer2::Plugin::Auth::Extensible> -- possibly not yet ported to Dancer2

=item *

L<Dancer2::Plugin::Auth::RBAC> -- possibly not yet ported to Dancer2

=back

For password authentication algorithms for your own '/login' handler, see:

=over 4

=item *

L<Auth::Passphrase>

=item *

L<Dancer2::Plugin::Passphrase> -- possibly not yet ported to Dancer2

=back

=head1 ACKNOWLEDGMENTS

This simplified Auth module was inspired by L<Dancer::Plugin::Auth::Extensible>
by David Precious and discussions about its API by member of the Dancer Users
mailing list.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/dancer2-plugin-auth-tiny/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/dancer2-plugin-auth-tiny>

  git clone git://github.com/dagolden/dancer2-plugin-auth-tiny.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004



Changes
Revision history for Dancer2-Plugin-Auth-Tiny

{{$NEXT}}
    [CHANGED]

    - Update to Dancer2::Plugin 0.200000 (Plugin2).
      (Peter Mottram)

    [DOCUMENTED]

    - Correct "extend" example in documentation.
      (Biafra)

0.008     2016-01-23 18:20:25+01:00 Europe/Amsterdam

    [CHANGED]

    - Stop using deprecated plugin_args, removes warnings.
      (Peter Mottram)

0.007     2015-10-02 19:21:33+02:00 Europe/Amsterdam

    [DOCUMENTED]

    - Fix/update documentation for references.
      (Jason Crome, Jim Davis)

0.005     2014-11-24 00:58:46+01:00 Europe/Amsterdam

    [CHANGED]

    - Revert fix since Dancer2 0.154000 fixed it itself.

0.004     2014-11-03 19:41:23+01:00 Europe/Amsterdam

    [CHANGED]

    - Updated to support latest Dancer2 (now requires 0.153000)

    - Fixed a bug when path creation for mounted applications

0.003     2013-03-06 19:16:46 America/New_York

    [CHANGED]

    - Forked into Dancer2::Plugin::Auth::Tiny for proper Dancer2 support

    [DOCUMENTED]

    - fixed a typo in an example

***** Prior history refers to Dancer::Plugin::Auth::Tiny *****

0.002     2012-12-20 11:16:59 America/New_York

    [ADDED]

    - The 'passthrough' feature allows some query parameters in the
      original request to be passed through to the login request

0.001     2012-12-18 22:44:27 America/New_York

    - First release



Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Require logged-in user for specified routes",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>, Sawyer X <xsawyerx\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Dancer2-Plugin-Auth-Tiny",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Dancer2::Plugin::Auth::Tiny",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dancer2" => "0.200000",
    "Dancer2::Core::Types" => 0,
    "Dancer2::Plugin" => "0.200000",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => "0.200000",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => "0.19",
    "HTTP::Cookies" => 0,
    "HTTP::Request" => 0,
    "HTTP::Request::Common" => 0,
    "Plack::Builder" => 0,
    "Plack::Request" => 0,
    "Plack::Test" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.009",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dancer2" => "0.200000",
  "Dancer2::Core::Types" => 0,
  "Dancer2::Plugin" => "0.200000",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => "0.19",
  "HTTP::Cookies" => 0,
  "HTTP::Request" => 0,
  "HTTP::Request::Common" => 0,
  "Plack::Builder" => 0,
  "Plack::Request" => 0,
  "Plack::Test" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Dancer2/Plugin/Auth/Tiny.pm
use 5.008001;
use strict;
use warnings;

package Dancer2::Plugin::Auth::Tiny;
# ABSTRACT: Require logged-in user for specified routes
our $VERSION = '0.009';

use Carp qw/croak/;
use Dancer2::Core::Types qw/ArrayRef Dancer2Prefix Str/;
use Dancer2::Plugin 0.200000;

my %dispatch = ( login => \&_build_login, );

plugin_keywords 'needs' => sub {
    my ( $plugin, $condition, @args ) = @_;

    my $builder = $dispatch{$condition};

    if ( ref $builder eq 'CODE' ) {
        return $builder->( $plugin, @args );
    }
    else {
        croak "Unknown authorization condition '$condition'";
    }
};

sub extend {
    my ( $class, @args ) = @_;
    unless ( @args % 2 == 0 ) {
        croak "arguments to $class\->extend must be key/value pairs";
    }
    %dispatch = ( %dispatch, @args );
}

sub _build_login {
    my ( $plugin, $coderef ) = @_;

    return sub {
        my $request = $plugin->app->request;
        if ( $plugin->app->session->read( $plugin->logged_in_key ) ) {
            goto $coderef;
        }
        else {
            my $params = $request->params;
            my $data =
              { $plugin->callback_key => $request->uri_for( $request->path, $params->{query} ) };
            for my $k ( @{ $plugin->passthrough } ) {
                $data->{$k} = $params->{$k} if $params->{$k};
            }
            return $plugin->app->redirect( $request->uri_for( $plugin->login_route, $data ) );
        }
    };
}

has login_route => (
    is => 'ro',
    isa => Dancer2Prefix,
    from_config => sub { '/login' },
);

has logged_in_key => (
    is => 'ro',
    isa => Str,
    from_config => sub { 'user' },
);

has callback_key => (
    is => 'ro',
    isa => Str,
    from_config => sub { 'return_url' },
);

has passthrough => (
    is => 'ro',
    isa => ArrayRef,
    from_config => sub { [qw/user/] },
);

1;

=for Pod::Coverage extend

=head1 SYNOPSIS

  use Dancer2::Plugin::Auth::Tiny;

  get '/private' => needs login => sub { ... };

  get '/login' => sub {
    # put 'return_url' in a hidden form field
    template 'login' => { return_url => params->{return_url} };
  };

  post '/login' => sub {
    if ( _is_valid( params->{user}, params->{password} ) ) {
      session user => params->{user};
      return redirect params->{return_url} || '/';
    }
    else {
      template 'login' => { error => "invalid username or password" };
    }
  };

  sub _is_valid { ... } # this is up to you

=head1 DESCRIPTION

This L<Dancer2> plugin provides an extremely simple way of requiring that a user
be logged in before allowing access to certain routes.

It is not "Tiny" in the usual CPAN sense, but it is "Tiny" with respect to
Dancer2 authentication plugins.  It provides very simple sugar to wrap route
handlers with an authentication closure.

The plugin provides the C<needs> keyword and a default C<login> wrapper that
you can use like this:

  get '/private' => needs login => $coderef;

The code above is roughly equivalent to this:

  get '/private' => sub {
    if ( session 'user' ) {
      goto $coderef;
    }
    else {
      return redirect uri_for( '/login',
        { return_url => uri_for( request->path, request->params ) } );
    }
  };

It is up to you to provide the '/login' route, handle actual authentication,
and set C<user> session variable if login is successful.

If the original request contains a parameter in the C<passthrough> list, it
will be added to the login query. For example,
C<http://example.com/private?user=dagolden> will be redirected as
C<http://example.com/login?user=dagolden&return_url=...>.  This facilitates
pre-populating a login form.

=head1 CONFIGURATION

You may override any of these settings:

=for :list
* C<login_route: /login> -- defines where a protected route is redirected
* C<logged_in_key: user> -- defines the session key that must be true to indicate a logged-in user
* C<callback_key: return_url> -- defines the parameter key with the original request URL that is passed to the login route
* C<passthrough: - user> -- a list of parameters that should be passed through to the login handler

=head1 EXTENDING

The class method C<extend> may be used to add (or override) authentication
criteria. For example, to add a check for the C<session 'is_admin'> key:

  Dancer2::Plugin::Auth::Tiny->extend(
    admin => sub {
      my ($plugin, $coderef) = @_;
      return sub {
        if ( $plugin->app->session->read("is_admin") ) {
          goto $coderef;
        }
        else {
          $plugin->app->redirect('/access_denied');
        }
      };
    }
  );

  get '/super_secret' => needs admin => sub { ... };

It takes key/value pairs where the value must be a closure generator that wraps
arguments passed to C<needs>.

You could pass additional arguments before the code reference like so:

  # don't conflict with Dancer2's any()
  use Syntax::Keyword::Junction 'any' => { -as => 'any_of' };

  Dancer2::Plugin::Auth::Tiny->extend(
    any_role => sub {
      my $coderef = pop;
      my ($plugin, @requested_roles) = @_;
      return sub {
        my @user_roles = @{ $plugin->app->session->read("roles") || [] };
        if ( any_of(@requested_roles) eq any_of(@user_roles) ) {
          goto $coderef;
        }
        else {
          $plugin->app->redirect '/access_denied';
        }
      };
    }
  );

  get '/parental' => needs any_role => qw/mom dad/ => sub { ... };

=head1 SEE ALSO

For more complex L<Dancer2> authentication, see:

=for :list
* L<Dancer2::Plugin::Auth::Extensible>
* L<Dancer2::Plugin::Auth::RBAC> -- possibly not yet (or going to be) ported to Dancer2
* L<Dancer2::Plugin::Auth::YARBAC>

For password authentication algorithms for your own '/login' handler, see:

=for :list
* L<Auth::Passphrase>
* L<Dancer2::Plugin::Passphrase>

=head1 ACKNOWLEDGMENTS

This simplified Auth module was inspired by L<Dancer::Plugin::Auth::Extensible>
by David Precious and discussions about its API by members of the Dancer Users
mailing list.

=for Pod::Coverage ClassHooks PluginKeyword dancer_app execute_plugin_hook hook on_plugin_import plugin_args plugin_setting register register_hook register_plugin request var

=cut

# vim: ts=4 sts=4 sw=4 et:
